nameresolution/sqlite/sqlite.go

```diff
@@ -0,0 +1,317 @@
+	go s.renewRegistration()
+
+	return nil
+}
+
+func (s *resolver) initGC() (err error) {
+	s.gc, err = internalsql.ScheduleGarbageCollector(internalsql.GCOptions{
+		Logger: s.logger,
+		UpdateLastCleanupQuery: func(arg any) (string, any) {
+			return fmt.Sprintf(`INSERT INTO %s (key, value)
+				VALUES ('nr-last-cleanup', CURRENT_TIMESTAMP)
+				ON CONFLICT (key)
+				DO UPDATE SET value = CURRENT_TIMESTAMP
+					WHERE unixepoch(CURRENT_TIMESTAMP) - unixepoch(value)  > ?;`,
+				s.metadata.MetadataTableName,
+			), arg
+		},
+		DeleteExpiredValuesQuery: fmt.Sprintf(
+			`DELETE FROM %s WHERE unixepoch(CURRENT_TIMESTAMP) - last_update < %d`,
+			s.metadata.TableName,
+			int(s.metadata.UpdateInterval.Seconds()),
+		),
+		CleanupInterval: s.metadata.CleanupInterval,
+		DB:              internalsql.AdaptDatabaseSQLConn(s.db),
+	})
+	return err
+}
+
+// Registers the host
+func (s *resolver) registerHost(ctx context.Context) error {
+	// Get the registration ID
+	u, err := uuid.NewRandom()
+	if err != nil {
+		return fmt.Errorf("failed to generate registration ID: %w", err)
+	}
+	s.registrationID = u.String()
+
+	queryCtx, queryCancel := context.WithTimeout(ctx, s.metadata.Timeout)
+	defer queryCancel()
+
+	// There's a unique index on address
+	// We use REPLACE to take over any previous registration for that address
+	// TODO: Add support for namespacing. See https://github.com/dapr/components-contrib/issues/3179
+	_, err = s.db.ExecContext(queryCtx,
+		fmt.Sprintf("REPLACE INTO %s (registration_id, address, app_id, namespace, last_update) VALUES (?, ?, ?, ?, unixepoch(CURRENT_TIMESTAMP))", s.metadata.TableName),
+		s.registrationID, s.metadata.GetAddress(), s.metadata.appID, "",
+	)
+	if err != nil {
+		return fmt.Errorf("failed to register host: %w", err)
+	}
+
+	return nil
+}
+
+// In backgrounds, periodically renews the host's registration
+// Should be invoked in a background goroutine
+func (s *resolver) renewRegistration() {
+	defer s.wg.Done()
+
+	addr := s.metadata.GetAddress()
+
+	d := s.metadata.UpdateInterval - s.metadata.Timeout
+	s.logger.Debugf("Started renewing host registration in background with interval %v", s.metadata.UpdateInterval)
+	t := time.NewTicker(d)
+	defer t.Stop()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	for {
+		select {
+		case <-s.closeCh:
+			// Component is closing
+			s.logger.Debug("Stopped renewing host registration: component is closing")
+			return
+
+		case <-t.C:
+			// Renew on the ticker
+			go func() {
+				err := s.doRenewRegistration(ctx, addr)
+				if err != nil {
+					// Log errors
+					s.logger.Errorf("Failed to update host registration: %v", err)
+
+					if errors.Is(err, errRegistrationLost) {
+						// This means that our registration has been taken over by another host
+						// It should never happen unless there's something really bad going on
+						// Panicking here to force a restart of Dapr
+						s.logger.Fatalf("Host registration lost")
```

Comment:
Please don't panic. All active connections and resources will not be given the time to correctly gracefully close. We should shutdown the component instead.

Reply:
Ok

Reply:
Looking into it. I am not sure what the best way to cascade this error is in this case, since this runs on a background goroutine that has no way to communicate the error to a caller.

I am concerned that just calling `Close()` on the component could put the sidecar in an un-healthy state where all service invocation wouldn't work, but the sidecar would still be reported as healthy and so there'd be no way for the platform (like K8s or Docker) to detect the failure and restart.

Do you have any suggestion?

Reply:
Ideally, since there is always at most 1 name resolver we probably should have a single long running `Run` style function so that we can correctly propagate these errors back to runtime and shutdown appropriately. Are we in a position to change the nameresolver interface given that it is only imported by runtime and not a component, or do we need to do "v2"? Conversely, we can also create a new long running `healthz` type function which does the same.

I agree with you that since the name resolver is a special case "component", we need to propagate the error back and not just `Close`.

Reply:
> Are we in a position to change the nameresolver interface given that it is only imported by runtime and not a component, or do we need to do "v2"?

Yes, the APIs of components-contrib are not stable (meaning, we don't have a stable contract with third-party users).

Of course, if this is what we want to do, it's something that should not be done in this PR :)

> Conversely, we can also create a new long running healthz type function which does the same.

I am not sure if this issue with possible conflicts would be unique to this component. I need to look at the other implementations.

> I agree with you that since the name resolver is a special case "component", we need to propagate the error back and not just Close.

Let's resolve this for now and accept the panic as a temporary workaround? We can address the issue more properly with API changes in the runtime in a separate PR.